plenum/server/replica.py

```diff
@@ -292,6 +292,15 @@ def __init__(self, node: 'plenum.server.node.Node', instId: int,
         # instance is
         self._primaryName = None  # type: Optional[str]
 
+        # PRE-PREPAREs timestamps stored by non primary replica to check
+        # obsolescence of incoming PrePrepares. Pre-prepares with the same
+        # 3PC key are not merged since we need to keep incoming timestamps
+        # for each new PP from every nodes separately.
+        # Dictionary:
+        #   key: Tuple[pp.viewNo, pp.seqNo]
+        #   value: Dict[PrePrepare, Tuple[sender, timestamp]]
+        self.pre_prepare_tss = defaultdict(dict)
```

Comment:
Why do we need to have a dict of dict and store pre-prepares there?
I think just a simple Dict[(pp.viewNo, pp.seqNo), timestamp] is enough, isn't it?
If this is because of duplicate PrePrepares for the same 3PC key, then this is invalid PrePrepare, which is validated with PP_CHECK_DUPLICATE raised.

Reply:
Dict[(pp.viewNo, pp.seqNo), timestamp] won't work since:
- the same PrePrepare might come multiple times (e.g. for PP that includes not finalized requests, or missed previous PP or with bad timestamps), but we need to set timestamp only first time
- (pp.viewNo, pp.seqNo) might be the same for multiple different PPs (different senders, or single malicious one)